File: plots/revenue_trend.py
from ppt_gen.helpers import finalize_line_chart
from ppt_gen.registry import register
from ppt_gen.slide_context import SlideContext


@register("revenue_trend")
def render(ctx: SlideContext):
    fig, ax = ctx.figure()
    df = ctx.data.get("quarterly")
    if df is None:
        months = ["Jan", "Feb", "Mar", "Apr", "May", "Jun"]
        revenue = [4.2, 4.5, 4.8, 5.1, 5.4, 5.9]
    else:
        months = df["month"].tolist()
        revenue = df["revenue_m"].tolist()

    ax.plot(
        months,
        revenue,
        label="Revenue",
        color=ctx.color(0),
        marker="o",
        linewidth=ctx.tokens.plot["linewidth"],
        markersize=ctx.tokens.plot["markersize"],
    )
    ax.set_title("Revenue trend")
    ax.set_xlabel("Month")
    ax.set_ylabel("Revenue (M)")
    ax.grid(True, alpha=ctx.tokens.plot["grid_alpha"])
    finalize_line_chart(fig, ax, ctx.tokens, legend_position="upper-left")
    return fig

File: charts/revenue_trend.mjs
import { register } from "../lib/charts/registry.mjs";
import { renderChartPng } from "../lib/charts/helpers.mjs";

register("revenue_trend", async (ctx) => {
  const rows = ctx.data.quarterly ?? [
    { month: "Jan", revenue_m: 4.2 },
    { month: "Feb", revenue_m: 4.5 },
    { month: "Mar", revenue_m: 4.8 },
    { month: "Apr", revenue_m: 5.1 },
    { month: "May", revenue_m: 5.4 },
    { month: "Jun", revenue_m: 5.9 },
  ];
  const months = rows.map((r) => r.month);
  const revenue = rows.map((r) => Number(r.revenue_m));
  const color = ctx.color(0);
  const lw = ctx.tokens.plot.linewidth ?? 2;
  const ms = ctx.tokens.plot.markersize ?? 6;

  return renderChartPng(ctx, () => ({
    type: "line",
    data: {
      labels: months,
      datasets: [
        {
          label: "Revenue",
          data: revenue,
          borderColor: color,
          backgroundColor: color,
          borderWidth: lw,
          pointRadius: ms / 2,
          pointHoverRadius: ms / 2 + 1,
          tension: 0.15,
        },
      ],
    },
    options: {
      plugins: {
        title: { display: true, text: "Revenue trend" },
        legend: { position: "top", align: "start" },
      },
      scales: {
        x: {
          title: { display: true, text: "Month" },
        },
        y: {
          title: { display: true, text: "Revenue (M)" },
          grace: "12%",
        },
      },
    },
  }));
});
